NDTensors/src/fill.jl

```diff
@@ -1,14 +1,54 @@
 # This is a file to write generic fills for NDTensors.
 #  This includes random fills, zeros, ...
 
+function generic_randn(
+  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer=0
+) where {DataT<:AbstractArray{ElT}} where {ElT}
+  data = generic_randn(DataT, dim)
+  return StoreT(data)
+end
+
+function generic_randn(StoreT::Type{<:Dense}, dim::Integer=0)
+  return generic_randn(StoreT{default_eltype(),default_datatype(default_eltype())}, dim)
+end
+
+function generic_randn(DataT::Type{<:AbstractArray{ElT}}, dim::Integer=0) where {ElT}
+  data = DataT(undef, dim)
+  for i in 1:dim
+    @inbounds data[i] = randn(ElT)
+  end
+  return data
+end
+
+function generic_randn(DataT::Type{<:AbstractArray}, dim::Integer=0)
+  return generic_randn(DataT{default_eltype()}, dim)
+end
+
+##TODO replace randn in ITensors with generic_randn
+## and replace zeros with generic_zeros
 function randn(
   StoreT::Type{<:Dense{ElT,VecT}}, dim::Integer
 ) where {VecT<:AbstractArray{ElT}} where {ElT<:Number}
   return StoreT(randn(ElT, dim))
 end
 
-function zeros(datatype::Type{<:AbstractArray{ElT}}, dim::Integer=0) where {ElT<:Number}
-  return fill!(datatype(undef, dim), zero(ElT))
+function generic_zeros(
+  StoreT::Type{<:Dense{ElT,DataT}}, dim::Integer=0
+) where {DataT<:AbstractArray{ElT}} where {ElT}
```

Comment:
Don't think you need ElT and `where {ElT}` here and can just use `<:AbstractArray`

Reply:
Or just drop `<:AbstractArray` since that is a constraint already put on the `Dense` type itself.

Reply:
So I tried rewriting this as 
```
  function generic_zeros( StoreT::Type{<:Dense{Any, DataT}}, dim::Integer = 0) where {DataT<:AbstractArray{Any}}
  ``` 
But you cannot access this function, if you call say `generic_zeros(Dense{Float32, Vector{Float32}}, 20)` the only function which is called is 
 ```
function generic_zeros(StoreT::Type{<:Dense}, dim::Integer = 0)
```
And fails. It would be possible to make a function. 
```
function generic_zeros(StoreT::Type{<:Dense{DataT}}, dim::Integer = 0) where {DataT<:AbstractArray}
  data = default_zeros(DataT{default_eltype()}, dim)
  return Dense(data)
end
```
But obviously the provided argument is not a "proper" dense type

Reply:
What about:
```julia
function generic_zeros(
  StoreT::Type{<:Dense{<:Any,DataT}}, dim::Integer=0
) where {DataT}
  [...]
end
```
`Dense{Any,DataT}` is taken literally as a `Dense` with an element type of `Any`.

Reply:
Oh, yes `<:Any` worked. I thought I was doing something wrong. Thanks!

Reply:
Hmm actually there was another problem with `<:Any` It can easily get passed by for a less specific constructor (i.e. it can fall into a more generic constructor) but it seems to work better if I specify `<:Number` instead of `<:Any`

Reply:
What's the particular issue, i.e. what inputs and then what function accidentally gets called? That seems strange and may indicates another function has a `<:Number` constraint that should be removed.